json: update json doc (#11180)
pingcap/docs-cn: data-type-json.md

SELECT id FROM city WHERE population >= 100;
```

# 使用限制

- TiDB 暂不支持 JSON PATH 中范围选取的语法，以下 SQL 语句会在 TiDB 中报错:

    ```sql
    SELECT j->'$[1 to 2]' FROM t;
    SELECT j->'$[last]' FROM t;
    ```

> - 请勿使用 BR 工具将备份的 `>=6.3` 版本包含 JSON 列的数据恢复至 `<6.3` 版本的集群中，否则有几率出现无法查询 JSON 的问题。

Reviewer: ```suggestion
- TiDB Backup & Restore（BR）在 v6.3.0 之前不支持恢复包含 JSON 列的数据。另外，任何版本的 BR 都不支持恢复包含 JSON 列的数据到 v6.3.0 之前的 TiDB 集群。
```

Reviewer: 麻烦参考 https://docs.pingcap.com/zh/tidb/stable/character-set-gbk#%E7%BB%84%E4%BB%B6%E5%85%BC%E5%AE%B9%E6%80%A7 看下是否要这么改

# MySQL 兼容性

- 当使用二进制类型数据创建 JSON 时，目前 MySQL 会将其误标记为 STRING 类型。

    ```sql
    CREATE TABLE test(a json);
    INSERT INTO test SELECT json_objectagg('a', b'01010101');

>     -- TiDB 结果如下所示, 在 MySQL 中下面 SQL 结果为 `0, 1`。

Reviewer: ```suggestion
    -- 在 TiDB 中，执行以下 SQL 语句返回结果如下所示。在 MySQL 中，执行以下 SQL 语句的结果为 `0, 1`。
```

    mysql> SELECT JSON_EXTRACT(JSON_OBJECT('a', b'01010101'), '$.a') = "base64:type15:VQ==" AS r1, JSON_EXTRACT(a, '$.a') = "base64:type15:VQ==" AS r2 FROM test;
    +------+------+
    | r1   | r2   |
    +------+------+
    |    0 |    0 |
    +------+------+
    1 row in set (0.01 sec)
    ```

>     详情可见此 [issue](https://github.com/pingcap/tidb/issues/37443)

Reviewer: ```suggestion
    详情可见此 [issue](https://github.com/pingcap/tidb/issues/37443)。
```

> - 当 `enum/set` 转换为 JSON 时，TiDB 会检查其格式正确性。下面 SQL 在 TiDB 中会报错:

Reviewer: ```suggestion
- 当 `enum/set` 转换为 JSON 时，TiDB 会检查其格式正确性。例如，当执行下面的 SQL 语句时，TiDB 中会报错：
```

    ```sql
    CREATE TABLE t(e ENUM('a'));
    INSERT INTO t VALUES ('a');
    mysql> SELECT CAST(e AS JSON) FROM t;
    ERROR 3140 (22032): Invalid JSON text: The document root must not be followed by other values.
    ```

>     详情可见此 [issue](https://github.com/pingcap/tidb/issues/9999)

Reviewer: ```suggestion
    详情可见此 [issue](https://github.com/pingcap/tidb/issues/9999)。
```

- TiDB 支持使用 `ORDER BY` 对 JSON Array 或 JSON Object 进行排序。

>     当使用 `ORDER BY` 对 JSON Array 或 JSON Object 进行排序时，MySQL 会返回一个警告，且排序结果与比较运算结果不一致:

Reviewer: ```suggestion
    当使用 `ORDER BY` 对 JSON Array 或 JSON Object 进行排序时，MySQL 会返回一个警告，且排序结果与比较运算结果不一致：
```

    ```sql
    CREATE TABLE t(j JSON);
    INSERT INTO t VALUES ('[1,2,3,4]');
    INSERT INTO t VALUES ('[5]');

    mysql> SELECT j FROM t WHERE j < JSON_ARRAY(5);
    +--------------+
    | j            |
    +--------------+
    | [1, 2, 3, 4] |
    +--------------+
    1 row in set (0.00 sec)

>     -- 在 MySQL 中，执行以下 SQL 语句会返回警告 “This version of MySQL doesn't yet support 'sorting of non-scalar JSON values'. ”，且排序结果与 `<` 比较结果不一致。

Reviewer: ```suggestion
    -- 在 TiDB 中，执行以下 SQL 语句返回结果如下所示。在 MySQL 中，执行以下 SQL 语句会返回警告 “This version of MySQL doesn't yet support 'sorting of non-scalar JSON values'. ”，且排序结果与 `<` 比较结果不一致。
```

    mysql> SELECT j FROM t ORDER BY j;
    +--------------+
    | j            |
    +--------------+
    | [1, 2, 3, 4] |
    | [5]          |
    +--------------+
    2 rows in set (0.00 sec)
    ```

>     详情可见此 [issue](https://github.com/pingcap/tidb/issues/37506)

Reviewer: ```suggestion
    详情可见此 [issue](https://github.com/pingcap/tidb/issues/37506)。
```

- 在 INSERT JSON 列时，TiDB 会将值隐式转换为 JSON

    ```sql
    CREATE TABLE t(col JSON);

>     -- TiDB INSERT 成功，MySQL 会报错: Invalid JSON text

Reviewer: ```suggestion

    -- 在 TiDB 中，执行以下 INSERT 语句成功。在 MySQL 中，执行以下 INSERT 语句将返回 Invalid JSON text 错误。
```

SELECT id FROM city WHERE population >= 100;
```

# MySQL 兼容性

> - 修复了 MySQL 使用二进制类型数据创建 JSON 时错误的将其标记为 STRING TYPE 的 bug：

Reviewer 2: 还有一堆别的类型，比如 DATE，DATETIME，TIMESTAMP，在修了 :crying_cat_face:

> - 修复了 MySQL 使用二进制类型数据创建 JSON 时错误的将其标记为 STRING TYPE 的 bug：

Reviewer: 如果这个 bug 不是我们修复的。TiDB 本身就没有这个问题，建议这里换个说法👆。

> {{< copyable "sql" >}}

Reviewer: “{{< copyable "sql" >}}“ 不再需要了，文档站会自动渲染增加复制 icon。麻烦请删除。

```sql

> create table test (a json);

Reviewer: 请将本文中的所有 SQL 关键字改成大写的。

insert into test select json_objectagg('a', b'01010101');

mysql> select json_extract(json_object('a', b'01010101'), '$.a') = "base64:type15:VQ==" as result;
+--------+
| result |
+--------+
|      0 |
+--------+
1 row in set (0.02 sec)

-- 在 MySQL 中下面 SQL 结果为 1.
mysql> select json_extract(a, '$.a') = "base64:type15:VQ==" as result from test;
+--------+
| result |
+--------+
|      0 |
+--------+
1 row in set (0.02 sec)
```
详情可见此 [issue](https://github.com/pingcap/tidb/issues/37443)

- 修复了 MySQL 错误的将 `enum/set` 转换为 `json` 的 bug, 下面 SQL 在 TiDB 中会报错:

{{< copyable "sql" >}}

```sql
create table t(e enum('a'));
insert into t values ('a');
mysql> select cast(e as json) from t;
ERROR 3140 (22032): Invalid JSON text: The document root must not be followed by other values.
```
详情可见此 [issue](https://github.com/pingcap/tidb/issues/9999)

> - 支持对 JSON ARRAY/OBJECT 的 `order by`

Reviewer: ```suggestion
- TiDB 支持使用 `ORDER BY` 对 JSON Array 或 JSON Object 进行排序。
```

> 如果在 MySQL 中对 JSON ARRAY/OBJECT 使用 order by 会得到一个 warning，且结果与比较运算不一致:

Reviewer: ```suggestion
当使用 `ORDER BY` 对 JSON Array 或 JSON Object 进行排序时，MySQL 会返回一个警告，且排序结果与比较运算结果不一致:
```

{{< copyable "sql" >}}

```sql
create table t(j JSON);
insert into t values ('[1,2,3,4]');
insert into t values ('[5]');

mysql> select j from t where j < json_array(5);
+--------------+
| j            |
+--------------+
| [1, 2, 3, 4] |
+--------------+
1 row in set (0.00 sec)

> -- 在 MySQL 中下面 SQL 会抛出 warning: This version of MySQL doesn't yet support 'sorting of non-scalar JSON values'. 且结果与 `<` 比较结果不一致

Reviewer: ```suggestion
-- 在 MySQL 中，执行以下 SQL 语句会返回警告 “This version of MySQL doesn't yet support 'sorting of non-scalar JSON values'. ”，且排序结果与 `<` 比较结果不一致。
```

mysql> select j from t order by j;
+--------------+
| j            |
+--------------+
| [1, 2, 3, 4] |
| [5]          |
+--------------+
2 rows in set (0.00 sec)
```
详情可见此 [issue](https://github.com/pingcap/tidb/issues/37506)

> - 暂不支持 JSON PATH 中范围选取的语法，以下 SQL 会在 TiDB 中报错:

Reviewer: ```suggestion
- TiDB 暂不支持 JSON PATH 中范围选取的语法，以下 SQL 语句会在 TiDB 中报错:
```

- 当将 ENUM 或 SET 数据类型转换为 JSON 时，TiDB 会检查其格式正确性。例如，当执行下面的 SQL 语句时，TiDB 中会报错：

    ```sql
    CREATE TABLE t(e ENUM('a'));
    INSERT INTO t VALUES ('a');
    mysql> SELECT CAST(e AS JSON) FROM t;
    ERROR 3140 (22032): Invalid JSON text: The document root must not be followed by other values.
    ```

    详情可见此 [issue](https://github.com/pingcap/tidb/issues/9999)。

- TiDB 支持使用 `ORDER BY` 对 JSON Array 或 JSON Object 进行排序。

    当使用 `ORDER BY` 对 JSON Array 或 JSON Object 进行排序时，MySQL 会返回一个警告，且排序结果与比较运算结果不一致：

    ```sql
    CREATE TABLE t(j JSON);
    INSERT INTO t VALUES ('[1,2,3,4]');
    INSERT INTO t VALUES ('[5]');

    mysql> SELECT j FROM t WHERE j < JSON_ARRAY(5);
    +--------------+
    | j            |
    +--------------+
    | [1, 2, 3, 4] |
    +--------------+
    1 row in set (0.00 sec)

    -- 在 TiDB 中，执行以下 SQL 语句返回结果如下所示。在 MySQL 中，执行以下 SQL 语句会返回警告 “This version of MySQL doesn't yet support 'sorting of non-scalar JSON values'. ”，且排序结果与 `<` 比较结果不一致。
    mysql> SELECT j FROM t ORDER BY j;
    +--------------+
    | j            |
    +--------------+
    | [1, 2, 3, 4] |
    | [5]          |
    +--------------+
    2 rows in set (0.00 sec)

>     ```

Author: ```suggestion
    ```

```

SELECT id FROM city WHERE population >= 100;
```

# 使用限制

- 暂不支持将 JSON 函数下推至 TiFlash。
- TiDB 暂不支持 JSON PATH 中范围选取的语法，以下 SQL 语句会在 TiDB 中报错:

    ```sql
    SELECT j->'$[1 to 2]' FROM t;
    SELECT j->'$[last]' FROM t;
    ```

- TiDB Backup & Restore（BR）在 v6.3.0 之前不支持恢复包含 JSON 列的数据。另外，任何版本的 BR 都不支持恢复包含 JSON 列的数据到 v6.3.0 之前的 TiDB 集群。

> - 请勿使用任何同步工具同步非标准 JSON 类型（例如 DATE/DATETIME/TIME 等）的数据。

Author: 数据可以同步，但是类型会丢失，因为是 JSON 标准中没有这些类型。

> - 请勿使用任何同步工具同步非标准 JSON 类型（例如 DATE/DATETIME/TIME 等）的数据。

Reviewer: ```suggestion
- 请勿使用任何同步工具同步非标准 JSON 类型（例如 DATE、DATETIME、TIME 等）的数据。
```

## MySQL 兼容性

> - 当使用二进制类型数据创建 JSON 时，目前 MySQL 会将其误标记为 STRING 类型。

Reviewer 3: 这里是不是加一句，"而 TiDB 会维持正确的二进制类型"。

> - 当使用二进制类型数据创建 JSON 时，目前 MySQL 会将其误标记为 STRING 类型，而 TiDB 会维持正确的二进制类型。

Author: ```suggestion
- 当使用二进制类型数据创建 JSON 时，目前 MySQL 会将其误标记为 STRING 类型，而 TiDB 会保持正确的二进制类型。
```

    ```sql
    CREATE TABLE t(e ENUM('a'));
    INSERT INTO t VALUES ('a');
    mysql> SELECT CAST(e AS JSON) FROM t;
    ERROR 3140 (22032): Invalid JSON text: The document root must not be followed by other values.
    ```

    详情可见此 [issue](https://github.com/pingcap/tidb/issues/9999)。

- TiDB 支持使用 `ORDER BY` 对 JSON Array 或 JSON Object 进行排序。

    当使用 `ORDER BY` 对 JSON Array 或 JSON Object 进行排序时，MySQL 会返回一个警告，且排序结果与比较运算结果不一致：

    ```sql
    CREATE TABLE t(j JSON);
    INSERT INTO t VALUES ('[1,2,3,4]');
    INSERT INTO t VALUES ('[5]');

    mysql> SELECT j FROM t WHERE j < JSON_ARRAY(5);
    +--------------+
    | j            |
    +--------------+
    | [1, 2, 3, 4] |
    +--------------+
    1 row in set (0.00 sec)

    -- 在 TiDB 中，执行以下 SQL 语句返回结果如下所示。在 MySQL 中，执行以下 SQL 语句会返回警告 “This version of MySQL doesn't yet support 'sorting of non-scalar JSON values'. ”，且排序结果与 `<` 比较结果不一致。
    mysql> SELECT j FROM t ORDER BY j;
    +--------------+
    | j            |
    +--------------+
    | [1, 2, 3, 4] |
    | [5]          |
    +--------------+
    2 rows in set (0.00 sec)
    ```

    详情可见此 [issue](https://github.com/pingcap/tidb/issues/37506)。

> - 在 INSERT JSON 列时，TiDB 会将值隐式转换为 JSON

Reviewer 4: ```suggestion
- 在 INSERT JSON 列时，TiDB 会将值隐式转换为 JSON：
```
